File: AndroidWebBox/coconut-network/src/main/kotlin/com/sniper/coconut/network/interceptors/LogInterceptor.kt
package com.sniper.coconut.network.interceptors

import com.sniper.coconut.network.HttpConfig
import com.sniper.coconut.network.HttpRequest
import com.sniper.coconut.network.HttpResponse
import com.sniper.coconut.network.utils.Logger
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

private const val TAG = "HttpClient"

/**
 * 日志拦截器
 * 记录请求和响应信息，敏感字段脱敏
 */
class LogInterceptor(config: HttpConfig? = null) : RequestInterceptor {
    private val sensitiveHeaders: Set<String> =
        (config?.sensitiveHeaders ?: listOf("Authorization", "Cookie"))
            .flatMap { listOf(it.lowercase(), it) }.toSet()

    private val sensitiveParams: Set<String> =
        (config?.sensitiveParams ?: listOf("password", "token", "secret"))
            .flatMap { listOf(it.lowercase(), it) }.toSet()

    override suspend fun onRequest(request: HttpRequest): HttpRequest {
        val maskedHeaders = maskHeaders(request.headers)
        val maskedParams = maskParams(request.params)
        val bodyStr = request.body?.let { maskBody(it) } ?: "null"

        Logger.i(TAG, "--> ${request.method} ${request.url}")
        Logger.i(TAG, "    Headers: $maskedHeaders")
        Logger.i(TAG, "    Params: $maskedParams")
        Logger.i(TAG, "    Body: $bodyStr")

        return request
    }

    override suspend fun onResponse(response: HttpResponse): HttpResponse {
        Logger.i(TAG, "<-- HTTP ${response.httpStatus} code=${response.code} cost=${response.costTime}ms")
        Logger.i(TAG, "    Msg: ${response.msg}")

        return response
    }

    /** 脱敏 Header */
    private fun maskHeaders(headers: Map<String, String>): String {
        val result = LinkedHashMap<String, String>()
        for ((key, value) in headers) {
            result[key] = if (key in sensitiveHeaders) "***" else value
        }
        return result.toString()
    }

    /** 脱敏查询参数 */
    private fun maskParams(params: Map<String, String>): String {
        val result = LinkedHashMap<String, String>()
        for ((key, value) in params) {
            result[key] = if (key in sensitiveParams) "***" else value
        }
        return result.toString()
    }

    /** 脱敏请求体 */
    private fun maskBody(body: JsonElement): String {
        return try {
            if (body is JsonObject) {
                val result = LinkedHashMap<String, JsonElement>()
                for ((key, value) in body) {
                    result[key] = if (key in sensitiveParams) JsonPrimitive("***") else value
                }
                result.toString()
            } else {
                body.toString()
            }
        } catch (e: Exception) {
            body.toString()
        }
    }
}

File: iOSWebBox/CoconutNetwork/Sources/CoconutNetwork/interceptors/LogInterceptor.swift
import Foundation

/// 日志拦截器 — 记录请求和响应信息，敏感字段脱敏。
public final class LogInterceptor: RequestInterceptor, @unchecked Sendable {

    private let sensitiveHeaders: Set<String>
    private let sensitiveParams: Set<String>

    public init(config: HttpConfig? = nil) {
        let headers = config?.sensitiveHeaders ?? ["Authorization", "Cookie"]
        let params = config?.sensitiveParams ?? ["password", "token", "secret"]
        self.sensitiveHeaders = Set(headers.flatMap { [$0.lowercased(), $0] })
        self.sensitiveParams = Set(params.flatMap { [$0.lowercased(), $0] })
    }

    public func onRequest(_ request: HttpRequest) async -> HttpRequest {
        let maskedHeaders = Self.maskHeaders(request.headers, sensitive: sensitiveHeaders)
        let maskedParams = Self.maskParams(request.params, sensitive: sensitiveParams)
        let bodyStr = request.body.map { Self.maskBody($0, sensitive: sensitiveParams) } ?? "null"

        NetworkLog.i("HttpClient", "--> \(request.method.rawValue) \(request.url)")
        NetworkLog.i("HttpClient", "    Headers: \(maskedHeaders)")
        NetworkLog.i("HttpClient", "    Params: \(maskedParams)")
        NetworkLog.i("HttpClient", "    Body: \(bodyStr)")

        return request
    }

    public func onResponse(_ response: HttpResponse) async -> HttpResponse {
        NetworkLog.i("HttpClient", "<-- HTTP \(response.httpStatus) code=\(response.code) cost=\(response.costTime)ms")
        NetworkLog.i("HttpClient", "    Msg: \(response.msg)")
        return response
    }

    /// 脱敏 Header
    private static func maskHeaders(_ headers: [String: String], sensitive: Set<String>) -> String {
        var result: [String: String] = [:]
        for (key, value) in headers {
            result[key] = sensitive.contains(key) ? "***" : value
        }
        return "\(result)"
    }

    /// 脱敏查询参数
    private static func maskParams(_ params: [String: String], sensitive: Set<String>) -> String {
        var result: [String: String] = [:]
        for (key, value) in params {
            result[key] = sensitive.contains(key) ? "***" : value
        }
        return "\(result)"
    }

    /// 脱敏请求体
    private static func maskBody(_ body: JSONValue, sensitive: Set<String>) -> String {
        guard case .object(let object) = body else {
            return body.serializedString() ?? "\(body)"
        }
        var result: [String: JSONValue] = [:]
        for (key, value) in object {
            result[key] = sensitive.contains(key) ? .string("***") : value
        }
        return JSONValue.object(result).serializedString() ?? "\(body)"
    }
}
